src/plugins/visualizations/public/react_embeddable/visualize_embeddable.tsx

```diff
@@ -0,0 +1,534 @@
+import { VIS_EVENT_TO_TRIGGER } from '../embeddable';
+import { getCapabilities, getInspector, getUiActions, getUsageCollection } from '../services';
+import { ACTION_CONVERT_TO_LENS } from '../triggers';
+import { urlFor } from '../utils/saved_visualize_utils';
+import type { SerializedVis, Vis } from '../vis';
+import { createVisInstance } from './create_vis_instance';
+import { getExpressionRendererProps } from './get_expression_renderer_props';
+import { saveToLibrary } from './save_to_library';
+import { deserializeState, serializeState } from './state';
+import {
+  ExtraSavedObjectProperties,
+  VisualizeApi,
+  VisualizeOutputState,
+  VisualizeRuntimeState,
+  VisualizeSerializedState,
+  isVisualizeSavedObjectState,
+} from './types';
+
+export const getVisualizeEmbeddableFactory: (deps: {
+  embeddableStart: EmbeddableStart;
+  embeddableEnhancedStart?: EmbeddableEnhancedPluginStart;
+}) => ReactEmbeddableFactory<VisualizeSerializedState, VisualizeRuntimeState, VisualizeApi> = ({
+  embeddableStart,
+  embeddableEnhancedStart,
+}) => ({
+  type: VISUALIZE_EMBEDDABLE_TYPE,
+  deserializeState,
+  buildEmbeddable: async (initialState: unknown, buildApi, uuid, parentApi) => {
+    // Handle state transfer from legacy visualize editor, which uses the legacy visualize embeddable and doesn't
+    // produce a snapshot state. If buildEmbeddable is passed only a savedObjectId in the state, this means deserializeState
+    // was never run, and it needs to be invoked manually
+    const state = isVisualizeSavedObjectState(initialState)
+      ? await deserializeState({
+          rawState: initialState,
+        })
+      : (initialState as VisualizeRuntimeState);
+
+    // Initialize dynamic actions
+    const dynamicActionsApi = embeddableEnhancedStart?.initializeReactEmbeddableDynamicActions(
+      uuid,
+      () => titlesApi.panelTitle.getValue(),
+      state
+    );
+    // if it is provided, start the dynamic actions manager
+    const maybeStopDynamicActions = dynamicActionsApi?.startDynamicActions();
+
+    const { titlesApi, titleComparators, serializeTitles } = initializeTitles(state);
+
+    // Count renders; mostly used for testing.
+    const renderCount$ = new BehaviorSubject<number>(0);
+    const hasRendered$ = new BehaviorSubject<boolean>(false);
+
+    // Track vis data and initialize it into a vis instance
+    const serializedVis$ = new BehaviorSubject<SerializedVis>(state.serializedVis);
+    const initialVisInstance = await createVisInstance(state.serializedVis);
+    const vis$ = new BehaviorSubject<Vis>(initialVisInstance);
+
+    // Track UI state
+    const onUiStateChange = () => serializedVis$.next(vis$.getValue().serialize());
+    initialVisInstance.uiState.on('change', onUiStateChange);
+    vis$.subscribe((vis) => vis.uiState.on('change', onUiStateChange));
+
+    // When the serialized vis changes, update the vis instance
+    serializedVis$.subscribe(async (serializedVis) => {
```

Comment:
This should use a switchMap to fetch async items. Then subscribe should apply them

```
serializedVis$.pipe(
  switchMap((serializedVis) => {
    const vis = await createVisInstance(serializedVis)
    const const { params, abortController } = await getExpressionParams();
  })
).subscribe((vis, params, abortController) => {
  vis$.next(vis);
  if (params) expressionParams$.next(params);
  expressionAbortController$.next(abortController);
})

```

Reply:
Looks like you lost the below in the fix. Is that still needed?

```
const currentVis = vis$.getValue();
      if (currentVis) currentVis.uiState.off('change', onUiStateChange);
```

Reply:
Probably necessary yeah, might have memory leaks without